tracer/src/Datadog.Trace/Propagators/ParseUtility.cs

```diff
@@ -0,0 +1,226 @@
+// <copyright file="ParseUtility.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using Datadog.Trace.Headers;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Propagators
+{
+    internal class ParseUtility
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<ParseUtility>();
+
+#if NETCOREAPP
+        public static ulong ParseFromHexOrDefault(ReadOnlySpan<char> value)
+        {
+            if (ulong.TryParse(value, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var result))
+            {
+                return result;
+            }
+
+            return default;
+        }
+#else
+        public static ulong ParseFromHexOrDefault(string value)
+        {
+            try
+            {
+                return Convert.ToUInt64(value, 16);
+            }
+            catch
+            {
+                return default;
+            }
+        }
+#endif
+
+        public static ulong? ParseUInt64<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter getter, string headerName)
+            where TCarrierGetter : struct, ICarrierGetter<TCarrier>
+        {
+            var headerValues = getter.Get(carrier, headerName);
+            var hasValue = false;
+
+            if (headerValues is string[] stringValues)
+            {
+                // Checking string[] allows to avoid the enumerator allocation.
+                foreach (string? headerValue in stringValues)
+                {
+                    if (ulong.TryParse(headerValue, NumberStyles.Integer, CultureInfo.InvariantCulture, out var result))
+                    {
+                        return result;
+                    }
+
+                    hasValue = true;
+                }
+            }
+            else if (TryParse(headerValues, ref hasValue, out var result))
+            {
+                return result;
+            }
+
+            if (hasValue)
+            {
+                Log.Warning(
+                    "Could not parse {HeaderName} headers: {HeaderValues}",
+                    headerName,
+                    string.Join(",", headerValues));
```

Comment:
This warning is new right? I wonder if this could cause an issue for customers if they have a bad service somewhere generating weird ids (e.g. out of range) with spamming the logs? Especially as we don't have log clean-up yet 🤔 

Not saying we _shouldn't_ log this, as it's obviously an important error, just that it seems like it's one that will either never be written, or will be written a _lot_. Might be safer to make it Debug until we've improved logs?

Reply:
nope, old bits: https://github.com/DataDog/dd-trace-dotnet/blob/master/tracer/src/Datadog.Trace/SpanContextPropagator.cs#L235-L241
Do you want me to change it here?

Reply:
I'm undecided 😄 I say leave it given it was already there